### ISSUE TITLE
Connection status callback not enough to tell connection to iothub is ready

### ISSUE THREAD

**Issue by alexmrtn** [Azure/azure-iot-sdk-c#1550]
Hi,

I have a question regarding the connection status callback that you can set using the `IoTHubDeviceClient_SetConnectionStatusCallback` function. I would use it to as event that tells the connection to the IoTHub is ready but it seems not sufficient.

As you can see in the log below I obtained from the `iothub_convenience_sample`, the message from the connection status callback is printed right after the connection to the iothub is established, but MQTT topic subscription is done afterward. 

```
Sending message 1 to IoTHub
Message: {"temperature":25.007,"humidity":79.143,"scale":"Celcius"}
-> 15:06:26 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: XXXXX.azure-devices.net/DDDDD/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.3.8%20(native%3b%20Linux%3b%20x86_64) | PWD: XXXX | CLEAN: 0
<- 15:06:26 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
The device client is connected to iothub // message from connection status callback
-> 15:06:26 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: devices/DDDDD/messages/devicebound/# | QOS: 1 | TOPIC_NAME: $iothub/methods/POST/# | QOS: 0
-> 15:06:26 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/DDDDD/messages/events/property_key=property_value&%24.cid=CORE_ID&%24.mid=MSG_ID&%24.ct=application%2fjson&%24.ce=utf-8 | PACKET_ID: 3 | PAYLOAD_LEN: 58
<- 15:06:26 SUBACK | PACKET_ID: 2 | RETURN_CODE: 1 | RETURN_CODE: 0
<- 15:06:26 PUBACK | PACKET_ID: 3
Confirmation callback received for message 1 with result IOTHUB_CLIENT_CONFIRMATION_OK
```

I would expect that the connection status is triggered not until all internal configuration is done ( mostly MQTT topic subscription for twin and telemetry). With the current implementation, I can not rely on the connection status to tell that my device is fully ready at this stage. 

Is there a way to be notified when all internal configuration is done? I'm thinking of having the connection status callback triggered after topic subscription, or extending the `IOTHUB_CLIENT_CONNECTION_STATUS_REASON` enum with a new value `IOTHUB_CLIENT_CONNECTION_READY`. What do you think?

Thanks, best regards.

**Comment by ericwol-msft** [comment]
There was a timing issue that was resolved with #1585 

Twin updates are placed on a queue, so that they are correctly handled on disconnect/reconnects. These queues have timeout, so that undelivered updates are notified to the application.

**Comment by Hmaalla** [comment]
For us, when connected status is received, that's means that all cloud/device synchronisation are finished and we can rely on this status to deal with twins , sending message etc
In the following example, we report some information on the twin just after receiving connected status, but when receiving complete device twin, we d'ont found the reported information

```
[DEBUG] -> 09:54:24 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: xxxx.azure-devices.net/xxxx/?api-version=2019-10-01&DeviceClientType=iothubclient%2f1.3.9%20(native%3b%20Linux%3b%20x86_64) | PWD: XXXX | CLEAN: 0
[DEBUG] <- 09:54:25 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
[INFO] Device Connected
[DEBUG] -> 09:54:25 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: devices/xxxxx/messages/devicebound/# | QOS: 1 | TOPIC_NAME: $iothub/twin/res/# | QOS: 0 | TOPIC_NAME: $iothub/methods/POST/# | QOS: 0
[DEBUG] <- 09:54:25 SUBACK | PACKET_ID: 2 | RETURN_CODE: 1 | RETURN_CODE: 0 | RETURN_CODE: 0
[DEBUG] -> 09:54:25 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/GET/?$rid=3
[DEBUG] -> 09:54:25 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/PATCH/properties/reported/?$rid=4 | PAYLOAD_LEN: 73
[DEBUG] -> 09:54:25 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/PATCH/properties/reported/?$rid=5 | PAYLOAD_LEN: 304
[DEBUG] -> 09:54:25 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/PATCH/properties/reported/?$rid=6 | PAYLOAD_LEN: 44
[DEBUG] <- 09:54:25 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: QOS_VALUE_INVALID | TOPIC_NAME: $iothub/twin/res/200/?$rid=3 | PAYLOAD_LEN: 424
[DEBUG] complet device twin is received (desired + reported)
[DEBUG] device twin full is received 
[DEBUG] desired: {} // empty desired
[DEBUG] reported: {} //reported is empty almost reporting is already done
````

**Comment by az-iot-builder-01** [comment]
@ericwol-msft, @Hmaalla, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
